pyrdfj2/j2_functions.py

```diff
@@ -0,0 +1,202 @@
+        content = re.sub(
+            triplus_pattern,  # find sequences of 3 or more quotes...
+            lambda x: esc_qt * len(x.group()),  # and have each of them escaped
+            content,  # so all ''' should become \'\'\' in the content
+        )
+    assert quote not in content, (
+        "ttl format error: still having "
+        f"applied quote format {quote} in text content"
+    )
+    return turtle_value(content, quote, "xsd:string", suffix)
+
+
+TTL_FMT_TYPE_FN = {
+    "xsd:boolean": turtle_format_boolean,
+    "xsd:integer": turtle_format_integer,
+    "xsd:double": turtle_format_double,
+    "xsd:date": turtle_format_date,
+    "xsd:datetime": turtle_format_datetime,
+    "xsd:anyURI": turtle_format_uri,
+    "xsd:string": turtle_format_string,
+}
+
+
+def turtle_format(content, type_name: str, quote: str = "'"):
+    assert quote in "'\"", "ttl format only accepts ' or \" as valid quotes."
+    if content is None:
+        content = ""
+
+    suffix = None
+    if type_name.startswith("@"):
+        suffix = type_name
+        # assuming string content for further quoting rules
+        type_name = "xsd:string"
+
+    type_format_fn = TTL_FMT_TYPE_FN.get(type_name, None)
+    assert type_format_fn is not None, (
+        "type_name '%s' not supported." % type_name
+    )
+
+    return type_format_fn(content, quote, suffix)
+
+
+def uritexpand(template: str, context):
+    return URITemplate(template).expand(context)
+
+
+def regexreplace(find: str, replace: str, text: str):
+    return re.sub(find, replace, text)
+
+
+class ValueMapper:
+    def __init__(self):
+        self._map = dict()
+
+    def add(self, key, val):
+        if key in self._map:
+            assert val == self._map[key], (
+                f"duplicate key {key} with distinct"
+                " values not allowed to build map"
+            )
+        self._map[key] = val
+
+    def apply(
+        self, record: dict, origin_name: str, target_name: str, fallback=None
+    ) -> None:
+        assert (
+            target_name not in record
+        ), "applying map refuses to overwrite content already in record"
+        key = record[origin_name]
+        val = self._map.get(key, fallback)
+        record[target_name] = val
+
+
+def map_build(
+    it: Iterable, key_name: str, val_name: str = None, cached_as: str = None
+) -> ValueMapper:
+    assert key_name, "cannot build map without valid key-name"
+    # note: id val_name is None, we just map to the whole record
+    if cached_as is not None and cached_as in Functions._cache:
+        return Functions._cache[cached_as]
+    # else - make map
+    vmap = ValueMapper()
+    # - populate it
+    for item in it:
+        target = item[val_name] if val_name is not None else item
+        vmap.add(item[key_name], target)
+    # add it to the cache
+    if cached_as is not None:
+        Functions._cache[cached_as] = vmap
+    return vmap
+
+
+def xml_unparse(wrapper):
```

Comment:
this is tied to pysubyt using xmlasdict wrapper classes

renaming to `def as_str (value)`  seems a neat way out ?

Reply:
Should we just remove it?

Reply:
surely valid option imho
(again though we need to think about pysubyt history --> some templates are out there that might be using xml_unparse)

Reply:
I think when we update pysubyt to a usefull release we should update this. So there wont be the need to use this, no?